File: Top_100_Liked_Questions/102._Binary_Tree_Level_Order_Traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        # iterative level order traversal
        q = [(root, 0)]
        ans = []
        while q:
            node, lvl = q.pop(0)
            if not node:
                continue
                
            if len(ans) < lvl+1:
                ans.append([node.val])
            else:
                ans[lvl].append(node.val)
            q.append((node.left, lvl+1))
            q.append((node.right, lvl+1))
        
        return ans
    
    
    def levelOrder2(self, root: TreeNode) -> List[List[int]]:
        # recursive level order traversal
        ans = []
        def helper(node, lvl):
            if not node:
                return
            
            if len(ans) == lvl:
                ans.append([])
            
            ans[lvl].append(node.val)
            helper(node.left, lvl+1)
            helper(node.right, lvl+1)
        
        helper(root, 0)
        return ans

File: Top_100_Liked_Questions/102._Binary_Tree_Level_Order_Traversal/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
    // iterative inorder traversal
    let ans = [];
    let q = [[root, 0]];
    
    while (q.length) {
        let [node, lvl] = q.shift();
        
        if (!node)
            continue
        
        if (ans.length === lvl)
            ans.push([]);
        
        ans[lvl].push(node.val);
        
        q.push([node.left, lvl+1]);
        q.push([node.right, lvl+1]);
    }
    
    return ans;
}

var levelOrder2 = function(root) {
    // recursive inorder traversal
    let ans = [];
    let helper = function(node, lvl) {
        if (!node)
            return
        
        if (ans.length === lvl)
            ans.push([]);
        
        ans[lvl].push(node.val);
        
        helper(node.left, lvl+1);
        helper(node.right, lvl+1);
    }
    
    helper(root, 0);
    return ans;
};